### PR TITLE
UI enhancement to support Quickstarts as CRs

### PR DESCRIPTION
**Pull request by abhinandan13jan**
# Issue
Addresses https://issues.redhat.com/browse/ODC-4751
and part of https://issues.redhat.com/browse/ODC-4752

**Solution**
Adds UI support for migration of QuickStarts to CRD

**How to Test the PR**
The QuickStart CRD and the CRs are placed in 
`frontend/packages/console-app/components/quickstarts/data/mock/yamls`

Step 1: ~~Apply `quick-start-crd.yaml` and wait for sometime for the server to update~~ (API now contains the CRD)
Step 2: Apply a few CRs present in individual YAML files in same directory.
Step 3: Check for the `QuickStartTile` in `+Add` page
Step 4: Check for `/quickstart` URL for the `QuickStart` catalog
Step 5: Click on a quickstart and check the `DrawerPanel`

# Screenshot
qs-apply
qs-cata

# Implementation
- [x] Converting QuickStarts to YAML
- [x] Script for installing QuickStart CRD and YAMLS for testing
- [x]  Moving existing QuickStart JSON to mocks
- [x]  Updating QuickStart types as per latest CRD structure
- [x] Fetching QuickStarts from API in QuickStart Tile
- [x] Fetching QuickStart from API in Drawer
- [x]  Updating components and utils as per latest type
- [x] Cleanup QuickStart utils which point to JSON ( currently unused )
- [x]  Serverless quickstart YAMLs  updated as it is currently being changed -- https://issues.redhat.com/browse/ODC-5059
- [x] Update exiting tests

# Tests 
Quick Start test suites updated

# Browser conformance
Chrome

### CODE REVIEW

File: frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx [hunk 1]
```diff
@@ -46,9 +48,21 @@ const QuickStartDrawer: React.FC<QuickStartDrawerProps> = ({
...
+    namespaced: false,
+    isList: true,
+  });
+
+  React.useEffect(() => {
+    if (qsData && qsLoaded && !qsError) {
+      const index = qsData.findIndex((qs) => qs.metadata.name === activeQuickStartID);
+      index > -1 && setQuickStart(qsData[index]);
+    }
+  }, [activeQuickStartID, qsData, qsLoaded, qsError]);
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx:65]: Do not load the quick starts here directly. instead use `QuickStartsLoader` to load the quick starts.

File: frontend/packages/console-app/src/components/quick-starts/data/mocks/yamls/quick-start-crd.yaml [hunk 2]
```diff
@@ -0,0 +1,201 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/data/mocks/yamls/quick-start-crd.yaml:1]: We don't need the CRD yaml anymore. The console operator has been updated with new CRD after this PR merged - https://github.com/openshift/console-operator/pull/480

File: frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx [hunk 3]
```diff
@@ -12,17 +14,21 @@ const QuickStartsLoader: React.FC<QuickStartsLoaderProps> = ({ children }) => {
...
   const permissionChecks = React.useRef<{ [name: string]: boolean }>({});
+  const [qs, qsLoaded, qsError] = useK8sWatchResource<QuickStart[]>({
+    kind: referenceForModel(QuickStartModel),
+    isList: true,
+  });
 
   React.useEffect(() => {
-    setQuickStarts(getQuickStarts());
-  }, []);
+    if (qs && qsLoaded && !qsError) setQuickStarts(qs);
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx:23]: You don't need extra state for this. Just use whatever is returned from the `useK8sWatchResource` hook. No need for this effect.

File: frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts [hunk 4]
```diff
@@ -1,4 +1,6 @@
-export const sampleApplicationQuickStart = {
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts:1]: Why do we still need typescript files for quick starts if everything is converted to yamls?

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx (hunk 1 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx]: wrapped Drawer inside Loader

File: frontend/packages/console-app/src/components/quick-starts/data/mocks/yamls/quick-start-crd.yaml (hunk 2 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/data/mocks/yamls/quick-start-crd.yaml]: removed the CRD

File: frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx (hunk 3 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx]: removed it

File: frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts (hunk 4 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts:1]: retaining most as test data for QSCatalog and QS related components. Removed a couple

File: frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx [hunk 5]
```diff
@@ -37,19 +36,25 @@ type DispatchProps = {
...
   onClose,
+  quickStarts,
 }) => {
   const [modalOpen, setModalOpen] = React.useState<boolean>(false);
-  const quickStart = getQuickStartByName(activeQuickStartID);
   const [scrollDirection, handleScrollCallback] = useScrollDirection();
   const { t } = useTranslation();
 
+  const index = quickStarts.findIndex((qs) => qs.metadata.name === activeQuickStartID);
+  const quickStart = quickStarts[index];
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx:56]:
```suggestion
  const quickStart =  quickStarts.find((qs) => qs.metadata.name === activeQuickStartID);
```

Comment (rohitkrai03) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx:56]: Also wrap this in `useMemo`

File: frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts (hunk 4 above)
Comment (rohitkrai03) [reply frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts:1]: For tests I think keeping just one would be fine.

File: frontend/packages/console-app/src/components/quick-starts/data/quick-start-test-data.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { exploreServerlessQuickStart } from './mocks/json/explore-serverless-quickstart';
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/data/quick-start-test-data.ts:1]: This is not needed anymore.

File: frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx [hunk 7]
```diff
@@ -1,28 +1,29 @@
...
 };
 
 const QuickStartsLoader: React.FC<QuickStartsLoaderProps> = ({ children }) => {
-  const [quickStarts, setQuickStarts] = React.useState<QuickStart[]>([]);
+  const [quickStarts] = useK8sWatchResource<QuickStart[]>({
+    kind: referenceForModel(QuickStartModel),
+    isList: true,
+  });
   const [allowedQuickStarts, setAllowedQuickStarts] = React.useState<QuickStart[]>([]);
   const [loaded, setLoaded] = React.useState<boolean>(!(quickStarts.length > 0));
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx:18]: `useK8sWatchResource` hook returns a loaded variable to track loading of resource. use that.

File: frontend/public/components/app.jsx [hunk 8]
```diff
@@ -153,29 +154,33 @@ class App_ extends React.PureComponent {
...
-              onDrawerChange={this._onNotificationDrawerToggle}
-            >
-              <AppContents />
-            </ConnectedNotificationDrawer>
-          </Page>
-          <CloudShell />
-          <GuidedTour />
-          <ConsoleNotifier location="BannerBottom" />
-        </QuickStartDrawer>
+        <QuickStartsLoader>
```
Comment (rohitkrai03) [inline frontend/public/components/app.jsx:157]: Move `QuickStartsLoader` into `QuickStartsDrawer`. use an intermediate component in between if needed.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/public/components/app.jsx (hunk 8 above)
Comment (abhinandan13jan) [reply frontend/public/components/app.jsx]: moved inside drawer and separated the panel component

File: frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx (hunk 7 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/loader/QuickStartsLoader.tsx:18]: using quickStartsLoaded

File: frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx (hunk 5 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx]: updated, not using Memo as quickStarts now moved to props.

File: frontend/packages/console-app/src/components/quick-starts/data/quick-start-test-data.ts (hunk 6 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/data/quick-start-test-data.ts]: I've now kept only 3 quickStarts as I am checking them in the QuickStart Catalog for 3 concurrent quickStarts. removed rest.

File: frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts (hunk 4 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/data/mocks/json/sample-application-quickstart.ts]: I've now kept only 3 quickStarts as I am checking them in the QuickStart Catalog for 3 concurrent quickStarts. removed rest.

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx [hunk 9]
```diff
@@ -51,7 +51,7 @@ const QuickStartContent: React.FC<QuickStartContentProps> = ({
           tasks={tasks}
           conclusion={conclusion}
           allTaskStatuses={allTaskStatuses}
-          nextQuickStart={nextQuickStart}
+          nextQuickStart={nextQuickStart[0]}
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx:54]: `nextQuickStart` could also be undefined in some cases where a quick start doesn't define any next quick start. Better to add a null check.

File: frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx [hunk 10]
```diff
@@ -65,47 +50,22 @@ const QuickStartDrawer: React.FC<QuickStartDrawerProps> = ({
...
+        <DrawerContent
+          panelContent={
+            <QuickStartsLoader>
+              {(quickStarts) => (
+                <QuickStartPanelContent
+                  quickStarts={quickStarts}
+                  handleClose={handleClose}
+                  activeQuickStartID={activeQuickStartID}
+                />
+              )}
```
Comment (rohitkrai03) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx:65]: Move this into panelContent variable above.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx (hunk 9 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx]: added

File: frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx (hunk 10 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx]: updated

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@abhinandan13jan I don't see icons for any of the quick starts in the catalog.

**Comment by abhinandan13jan** [comment]
> @abhinandan13jan I don't see icons for any of the quick starts in the catalog.

@rohitkrai03 Updated the icons

**Review by rohitkrai03 — APPROVED** [review]
/approve

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+import { useTranslation } from 'react-i18next';
+import {
+  DrawerPanelContent,
+  DrawerPanelBody,
+  DrawerHead,
+  DrawerActions,
+  DrawerCloseButton,
+  Title,
+} from '@patternfly/react-core';
+import { QuickStart } from './utils/quick-start-types';
```
Comment (nemesis09) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:12]: nit: absolute import statements should be above others

**Comment by nemesis09** [comment]
retest

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx (hunk 11 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx]: updated

**Pull request merged**